API: nvim_exec: function to source multiline Vimscript ("anonymous :source")

in42 opened the pull request:
Make changes to the pull request #8765 by @Siddhant085 so that it can be merged.
Also make the changes requested in that pull request.

Fixes #8722  

Added a new function to source strings.

in42 committed:
API: nvim_source: save/restore script context #11159

Use a constant for the script id.

bfredl commented on src/nvim/ex_cmds2.c:
```diff
@@ -3014,6 +3014,44 @@ static FILE *fopen_noinh_readbin(char *filename)
+    .buf = cmd,
+    .offset = 0,
+  };
+  current_sctx.sc_sid = 999999;
```
probably should use a defined constant. there already are some special values in `globals.h`, should probably add it after `SID_API_CLIENT` (and set it to -9).

bfredl replied:
also I think we might need to save and restore `current_sctx` in case `nvim_source` is called by a script (unless it is automatically restored after every command, or something).

in42 replied:
I have saved and restored the script context and have also moved the script id to a constant. Do I need to also save `sourcing_name`, `sourcing_lnum` like I saw is being done in other places e.g. in `execute_env(..)` in `main.c`?

bfredl replied:
I think so, `do_source` does it.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -53,6 +54,14 @@
+{
+    try_start();
+    do_source_str((char_u *)command.data);
+    try_end(err);
```
`nvim_source` should live next to `nvim_command_output` perhaps. At least not before macros.

Also, 2-space indent.

Siddhant085 committed:
API: nvim_source

justinmk commented:
This doesn't work. For example, the following test fails:

```
nvim('source', 'let x1=1\nlet x2=2\nlet x3=3\n')
      eq(1, nvim('eval', 'x1'))                                            
      eq(2, nvim('eval', 'x2'))
      eq(3, nvim('eval', 'x3'))
```

Other problems (could wait until a followup PR, but at least mutlline input needs to work, otherwise `nvim_source` is useless).

- Handling of `got_int` (i.e. user hits CTRL-C)
- Error reporting
- What to do about `SourcePre`/`SourcePost`/`SourceCmd` ?
- As with `nvim_command_output`, it's likely we might want a way to capture the output. This could be controlled by a parameter.

It also had these problems, which I fixed in a local branch:

- Some tests were wrong/insufficient.
- Missing treatment in `get_scriptname()`, so e.g. `:verbose set {option}?` crashed.
- `get_str_line()` should return NULL on error

justinmk committed:
API: nvim_source: fix multiline input

- DOCMD_REPEAT is needed to source all lines of input.
- Fix ":verbose set {option}?" by handling SID_STR in get_scriptname().

closes #8722

justinmk commented:
Should we name this `nvim_exec` (in analogy with `execute()` and `nvim_exec_lua`)?

bfredl commented:
A good stress test would be to replace `helpers.source` to use `nvim_source` (or `nvim_execute` or whatever) internally in the test suite. Though a few test failures will be expected, as some tests do use the file name of the temporary file (those could be changed to a separate `helpers.source_file` function).

in42 committed:
API: nvim_source_output

- Similar to nvim_source but will capture the output
- Add meaningful VimL tracebacks for nvim_source
- Handle got_int
- Add error reporting

in42 commented:
* Have added tracebacks. If `nvim_source` is called from a script, `:source (no file) called at <filename>:<lineno>` will be printed, otherwise just `:source (no file)` will be printed. I have tested this manually, will add a test case for this.
* Have handled `got_int` in the `do_source_str` function. `got_int` also will be handled in `try_end` in the `nvim_source()` function.
* Have added error reporting in the `nvim_source` function.
* For capturing the output, have added the `nvim_source_output` function.

TODO:
* Should I execute `SourcePre`/`SourcePost`/`SourceCmd` for `do_source_str`? Currently those autocmds expect a filename.
* `:[range]source` should work like `nvim_source(...)`
* Reimplement `helpers.source` to use `nvim_source`

in42 commented:
@bfredl , @justinmk can you review the changes?

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -85,6 +86,54 @@ void api_vim_free_all_mem(void)
+{
+  try_start();
+  if (do_source_str(command.data, "nvim_source(..)") == FAIL) {
+    EMSG(_("Error in executing!"));
```
This error message is never shown (and shouldn't be needed either, the actual error will already have displayed a message).

in42 replied:
I have removed the error message.

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -85,6 +86,54 @@ void api_vim_free_all_mem(void)
+  msg_silent++;
+  capture_ga = &capture_local;
+  if (do_source_str(command.data, "nvim_source_output(..)") == FAIL) {
+    EMSG(_("Error in executing!"));
```
ditto

in42 replied:
I have removed the error message.

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -85,6 +86,54 @@ void api_vim_free_all_mem(void)
   map_free(String, handle_T)(namespace_ids);
 }
 
+void nvim_source(String command, Error *err) FUNC_API_SINCE(7)
```
Missing docstring. Doesn't need be long, just something like "execute a multiline block of ex commands from a string" and list the args.

`command` arg should be renamed though, at it is probably not a single command. Maybe `src` ?

in42 replied:
Have added the docstrings, and renamed the arguments.

justinmk commented on src/nvim/ex_cmds2.c:
```diff
@@ -3015,6 +3015,74 @@ static FILE *fopen_noinh_readbin(char *filename)
+  retval = OK;
+  if (got_int) {
+    EMSG(_(e_interr));
+  }
```
handling got_int here doesn't accomplish anything since by this point, the entire script already finished. Try this:

    :echo nvim_source_output("let a=0\nwhile 1\nlet a=1+get(g:,'a',0)\ncall append('$',[a])\nendwhile")
    ctrl-c

I'll remove this line and add a test.

justinmk commented on src/nvim/ex_cmds2.c:
```diff
@@ -3036,30 +3036,50 @@ static char_u *get_str_line(int c, void *cookie, int ident)
-                      DOCMD_VERBOSE|DOCMD_NOWAIT|DOCMD_REPEAT);
+  current_sctx.sc_lnum = save_sourcing_lnum;
+
+  int retval = FAIL;
```
why remove `retval = do_cmdline(...)` ?

justinmk committed:
API: rename nvim_source => nvim_exec

- Eliminate nvim_source_output(): add boolean `output` param to
  nvim_exec() instead.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -72,10 +73,70 @@ void api_vim_free_all_mem(void)
+/// @param src      Vimscript code
+/// @param output   Capture and return all (non-error, non-shell |:!|) output
+/// @param[out] err Error details (Vim error), if any
+String nvim_exec(String src, Boolean output, Error *err)
```
this function is almost identical to `nvim_command_output`. Try to avoid duplicating lots of code. (But, I will do so in this case)
